Enhance flatc build system

jmillan opened the pull request:
Fixes #1154

Do only generate them when the flatbuffers schemas change by using a proper Meson target.

jmillan committed:
Meson: properly generate C++ flatbuffer header files

Fixes #1154

Do only generate them when the flatbuffer schemas change

ibc commented on worker/Makefile:
```diff
@@ -209,17 +207,8 @@ libmediasoup-worker: setup flatc
-		$(FLATC) --cpp --cpp-field-case-style lower --reflect-names --scoped-enums \
-			-o $(FLATC_OUT_DIR) ./fbs/*
-
-flatc-clean:
```
No need for this anymore?

jmillan replied:
nop, we don't need a specific clear for flatc anymore. the cleanup is done for the regular 'clean'  target as the generated sources reside in the same build directory.

ibc commented on worker/meson.build:
```diff
@@ -277,6 +277,57 @@ executable(
+flatbuffers_cpp_out_dir = 'FBS'
+flatbuffers_cpp_install_dir = meson.current_source_dir() + '/include'
+
+flatc = find_program('flatc')
```
Where this this find_program() look for flatc binary? In the PATH? Because it shouldn't.

jmillan replied:
It's not in the path. flatc is a subproject of mediasoup. It finds it there as the subproject installation is a dependency for the target.

ibc replied:
What I would like to hear is that it just uses the binary provided by the flat subproject. If it uses whatever `flatc` in the `PATH` this means problems.

ibc replied:
So can't we just make it use `flatc` provided by the subproject and fail otherwise if not found?

jmillan replied:
> So can't we just make it use flatc provided by the subproject and fail otherwise if not found?

Not as per the docs. 

I've pushed a commit where we force the version given by our flatbuffers subproject.

ibc reviewed (commented):
So a question (because this happens to me in the past):

If I delete a source .fbs file because it's no longer needed (or I rename it), will it be also removed from the compiled fbs dir?

jmillan commented:
> If I delete a source .fbs file because it's no longer needed (or I rename it), will it be also removed from the compiled fbs dir?

It seems that Meson does not solve that for us. So if we do that, apart from renaming or deleting the corresponding source file name in the files list in meson.build, we should remove the old generated file from both the build and the install dir.

jmillan committed:
force the flatc version of the flatbuffers subproject

In order to make sure no other version is run.

There is no way, as per the docs, to force only searching the binary in
the subsprojects folder.

jmillan commented:
>> If I delete a source .fbs file because it's no longer needed (or I rename it), will it be also removed from the compiled fbs dir?

> It seems that Meson does not solve that for us. So if we do that, apart from renaming or deleting the corresponding source file name in the files list in meson.build, we should remove the old generated file from both the build and the install dir.

I honestly don't see a solution for this rather than go and remove the files from the build and install folders.

jmillan commented:
@ibc, feel free to give it a try in case you see any possible enhancement here.

jmillan committed:
flatc-clean target in order to clean include/FBS folder

jmillan commented:
> >> If I delete a source .fbs file because it's no longer needed (or I rename it), will it be also removed from the compiled fbs dir?
> 
> > It seems that Meson does not solve that for us. So if we do that, apart from renaming or deleting the corresponding source file name in the files list in meson.build, we should remove the old generated file from both the build and the install dir.
> 
> I honestly don't see a solution for this rather than go and remove the files from the build and install folders.

Fixed now. A `make clean` will make it.

ibc reviewed (changes requested):
flatc-clean must be added in .PHONY in Makefile.

ibc commented:
> > If I delete a source .fbs file because it's no longer needed (or I rename it), will it be also removed from the compiled fbs dir?
> 
> 
> 
> It seems that Meson does not solve that for us. So if we do that, apart from renaming or deleting the corresponding source file name in the files list in meson.build, we should remove the old generated file from both the build and the install dir. 

What are the build and the install dir? I assume the install dir is worker/include/FBS, is it? And not sure what the build dir is but I expect it not to be part of the repo (so gitignored), am I wrong?

jmillan committed:
Makefile: Add 'flatc-clean' target to .PHONY

jmillan commented:
> What are the build and the install dir? I assume the install dir is worker/include/FBS, is it? And not sure what the build dir is but I expect it not to be part of the repo (so gitignored), am I wrong?

Build dir is within the build folder: 
* ./out/Release/build/FBS
* ./out/Debug/build/FBS

Install dir is `include/FBS`.

ibc reviewed (approved):
Nice

jmillan committed:
Revert "force the flatc version of the flatbuffers subproject"

This reverts commit 2e67a31a8d18b8c887b8ba9db935669e98e249c4.

This is not needed since flatc subproject installation "forces" meson
lookups for 'flatc' to use that exact binary.

ibc commented:
So this doesn't work as expected right? https://github.com/versatica/mediasoup/commit/0147ef27240f72aa0c9902ee53b95adaf55d2779

I mean, there is literally no way in any Unix based OS to know which version a binary in the PATH has other than running it first with --version (in case it supports it) and parsing the random output that it generates. Honestly I didn't like that thing in the commit you have reverted. It just couldn't work.

So, can we just force Meson to run a binary included in a subproject and ignore PATH completely?

jmillan committed:
make generated FBS files a dependency

jmillan committed:
create a meson.build for 'fbs' directory

jmillan committed:
remove _generated suffix to generated fbs files

jmillan committed:
add missing meson.build file

jmillan commented:
> So, can we just force Meson to run a binary included in a subproject and ignore PATH completely?

As said, our build depends on the subproject flatbuffers which comes with the flatc binary. Such subproject **forces** future meson `find_binary('flatc')` to return the just generated binary. So we are 100% sure that we will always use our flatc in our subproject.

jmillan commented:
Summary of this PR, which I consider ready to be merge:

I'm added a `meson.build` file to `worker/fbs` file:
- It generates the corresponding .h files
- Worker build depends on it

As a bonus track, generated `.h` files where sufficed with `_generated` before and I've removed such suffix.

EDIT: header files are not set in `worker/include/` anymore, hence there is no need to make that specific clean. The headers are imported from the folder where they are built, simply.

jmillan committed:
reorder dependency
